pytorch_lightning/utilities/jsonargparse_utils.py

```diff
@@ -0,0 +1,150 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Type
+from jsonargparse import ArgumentParser, ActionConfigFile
+from pytorch_lightning import Trainer, LightningModule, LightningDataModule
+from pytorch_lightning.callbacks import Callback
+
+
+class LightningArgumentParser(ArgumentParser):
+    def __init__(self, *args, **kwargs):
+        """Initialize argument parser that supports configuration file input"""
+        super().__init__(*args, **kwargs)
+        self.add_argument('--cfg',
+            action=ActionConfigFile,
+            help='Path to a configuration file in json or yaml format.')
+
+
+    def add_trainer_args(
```

Comment:
The add_*_args methods should include a `blocklist` argument, a list of arguments to exclude. For example, I may want to hide `tpu_cores` from the CLI arguments if my code does not support TPUs

Reply:
I was thinking of this, but hadn't implemented it yet. But will do.

Reply:
jsonargparse now has a the option to skip some arguments. Though I haven't added yet to the trainer cli.

Reply:
I am unsure about this. The config options from the lightning trainer could be considered like a standard, so removing might not be a good idea. In your example, TPUs not supported, instead of hiding wouldn't it be better to extend Trainer, having `tpu_cores` with a docstring saying that it is not supported and if given a value fail with a not supported message?

Reply:
If I understood correctly what you are proposing, it would require duplicating the entire `Trainer` docstring for the `ConstrainedTrainer` subclass with only a few lines of it changed. That seems very cumbersome, especially if you want to have different `ConstrainedTrainer`s for different tasks/environments.

Currently, I am using the following hack to limit the arguments exposed:

https://github.com/carmocca/PyLaia/blob/c99ff97432892c97c43c16aaba253895682714ce/laia/common/arguments.py#L253-L286

Reply:
For a case like PyLaia there is already the `skip` option in [add_class_arguments](https://omni-us.github.io/jsonargparse/#jsonargparse.signatures.SignatureArguments.add_class_arguments). But I don't think PyLaia is a good example for this discussion since it is not designed to use the `LightningCLI` in this pull request. Maybe could you explain why you are skipping those 8 arguments to understand if in general it would also make sense to skip them when using `LightningCLI`.

Reply:
Taking your snippet: 

```python
from mymodels import MyModel
from pytorch_lightning.utilities.jsonargparse_utils import LightningCLI
LightningCLI(MyModel)
```
Imagine `MyModel` fails to work with AMP due to some unrelated PyTorch bug. To avoid people getting confused by whatever error is thrown, Im thinking about being able to do

```python
LightningCLI(MyModel, skip=["precision"])
# or maybe using the key prefix to avoid conflicts if we support skipping non-trainer args
LightningCLI(MyModel, skip=["trainer.precision"])
```

If i'm not mistaken, currently to achieve this, either you don't use `TrainerCLI` or you do the following:

```python
class MyArgumentParser(LightningArgumentParser):
    def add_trainer_args(...):
        self.add_class_arguments(trainer_class, skip=["precision"])

class MyLightningCLI(LightningCLI):
    def init_parser(...):
        self.parser = MyArgumentParser(...)

MyLightningCLI(MyModel)
```

Reply:
Adding skip is simple. Still my doubt is that by removing arguments from the trainer, the user does not know why this argument is missing. It is supposed to be a lightning trainer so you kind of expect them to be there. Instead of simply removing them it could be something like `LightningCLI(MyModel, unsupported={"precision": "unsupported due to bug in XXX"})`. The option wouldn't be removed, but shown with this message in the help or if attempted to be used.

Maybe I am just overthinking this.

Reply:
supporting `skip` and `unsupported` does not seem exclusive to me. Some may want to just hide it and others to be able to overwrite it.

To me, most CLI users do not know about the `Trainer` class under the hood and won't notice some arguments are missing.

> Maybe I am just overthinking this.

Maybe :smile:

Reply:
@carmocca is this a requirement or can we leave this as a future improvement?

Reply:
We can leave it for the future. I would rather not block the PR over this